### PR TITLE
[WIP] Healthcheck support for specifying IP addresses as networks

### PR DESCRIPTION
**Pull request by benagricola**
This implements some new functionality if the python 3 compatible ipaddress module is available (`py2-ipaddress`).

## Without `py2-ipaddress`
Passing /32 IPs only:

```
# ./healthcheck.py --ip=1.2.3.4 --ip=1.2.3.5 --ip=1.2.3.6 --no-ip-setup
INFO[healthcheck] send announces for UP state to ExaBGP
announce route 1.2.3.4/32 next-hop self med 100
announce route 1.2.3.5/32 next-hop self med 101
announce route 1.2.3.6/32 next-hop self med 102
```

Passing CIDR notation networks only:
```
# ./healthcheck.py --ip=1.2.3.0/30 --ip=1.2.3.4/30 --no-ip-setup
usage: healthcheck.py [-h] [--debug | --silent | --syslog-facility [FACILITY]
                      | --no-syslog] [--name NAME] [--config FILE]
                      [--pid FILE] [--user USER] [--group GROUP]
                      [--interval N] [--fast-interval N] [--timeout N]
                      [--rise N] [--fall N] [--disable FILE] [--command CMD]
                      [--next-hop IP] [--ip IP] [--no-ip-setup]
                      [--label LABEL] [--start-ip N] [--up-metric M]
                      [--down-metric M] [--disabled-metric M] [--increase M]
                      [--community COMMUNITY]
                      [--large-community LARGECOMMUNITY] [--as-path ASPATH]
                      [--withdraw-on-down] [--execute CMD] [--up-execute CMD]
                      [--down-execute CMD] [--disabled-execute CMD]
healthcheck.py: error: argument --ip: invalid IPAddress value: '1.2.3.0/30'
```
 _(note: same behaviour occurs with a mix of CIDR notation and IPs)_

## With `py2-ipaddress`
Passing /32 IPs only:

```
./healthcheck.py --ip=1.2.3.4 --ip=1.2.3.5 --ip=1.2.3.6 --no-ip-setup
INFO[healthcheck] send announces for UP state to ExaBGP
announce route 1.2.3.4/32 next-hop self med 100
announce route 1.2.3.5/32 next-hop self med 101
announce route 1.2.3.6/32 next-hop self med 102
```

Passing CIDR notation networks only:
```
# ./healthcheck.py --ip=1.2.3.0/30 --ip=1.2.3.4/30 --no-ip-setup
INFO[healthcheck] send announces for UP state to ExaBGP
announce route 1.2.3.0/30 next-hop self med 100
announce route 1.2.3.4/30 next-hop self med 101
```

Passing mixed /32 IPs and CIDR notation networks:
```
./healthcheck.py --ip=1.2.3.1 --ip=1.2.3.2 --ip=1.2.3.3 --ip=1.2.3.4/30 --ip=1.2.3.8/30 --no-ip-setup
INFO[healthcheck] send announces for UP state to ExaBGP
announce route 1.2.3.1/32 next-hop self med 100
announce route 1.2.3.2/32 next-hop self med 101
announce route 1.2.3.3/32 next-hop self med 102
announce route 1.2.3.4/30 next-hop self med 103
announce route 1.2.3.8/30 next-hop self med 104
```

Passing mixed /32 IPs and CIDR notation networks, using the `--deaggregate-networks` option:
```
# ./healthcheck.py --ip=1.2.3.1 --ip=1.2.3.2 --ip=1.2.3.3 --ip=1.2.3.4/30 --ip=1.2.3.8/30 --no-ip-setup --deaggregate-networks
INFO[healthcheck] send announces for UP state to ExaBGP
announce route 1.2.3.1/32 next-hop self med 100
announce route 1.2.3.2/32 next-hop self med 101
announce route 1.2.3.3/32 next-hop self med 102
announce route 1.2.3.4/32 next-hop self med 103
announce route 1.2.3.5/32 next-hop self med 104
announce route 1.2.3.6/32 next-hop self med 105
announce route 1.2.3.7/32 next-hop self med 106
announce route 1.2.3.8/32 next-hop self med 107
announce route 1.2.3.9/32 next-hop self med 108
announce route 1.2.3.10/32 next-hop self med 109
announce route 1.2.3.11/32 next-hop self med 110
```

Essentially this should make it easier to announce a block of IP addresses that fall within the same CIDR notation, and also advertise the whole CIDR block if not using `--deaggregate-networks`. One limitation is it's currently not possible to deaggregate some networks but not others.

Couple of things I'm not completely happy with right now:

- Use of `has_py3_ipaddress` to change the program flow
- Command line argument usage, should this be a completely separate option, `--network`, available if the correct module is installed?
- No tests & not tested on Python3 

Thoughts on the implementation, suggestions for improvements etc?

---
This change is [Reviewable](https://reviewable.io/reviews/exa-networks/exabgp/526)

### CODE REVIEW

**Comment by vincentbernat** [comment]
Why not providing the support for Python 2 too?

```python
from ipaddr import IPNetwork as ip_address
```

File: lib/exabgp/application/healthcheck.py [hunk 1]
```diff
@@ -152,9 +156,23 @@ def parse():
...
+        g.add_argument("--deaggregate-networks",
+                       dest="deaggregate_networks", action="store_true",
+                       help="Deaggregate Networks specified in --ip")
+    else:
+        g.add_argument("--ip", metavar='IP',
+                       type=ip_address, dest="ips", action="append",
+                       help="advertise this IP address")
+        # Hide this option since it cannot be used without py3_ipaddress
+        g.add_argument("--deaggregate-networks",
+                       dest="deaggregate_networks", action="store_false",
```
Comment (vincentbernat) [inline lib/exabgp/application/healthcheck.py]: `action="store_true"` (should set the default to `False`)

File: lib/exabgp/application/healthcheck.py [hunk 2]
```diff
@@ -556,6 +580,11 @@ def main():
         if options.ip_setup:
             setup_ips(options.ips, options.label)
         drop_privileges(options.user, options.group)
+
+        # Parse defined networks into a list of IPs for advertisement
+        if has_py3_ipaddress and options.deaggregate_networks:
+            options.ips = [ip_address(ip) for net in options.ips for ip in net]
```
Comment (vincentbernat) [inline lib/exabgp/application/healthcheck.py]: Then, here, no need to test for `has_py3_ipaddress`.

File: lib/exabgp/application/healthcheck.py [hunk 3]
```diff
@@ -56,16 +56,20 @@
 
 logger = logging.getLogger("healthcheck")
 
+has_py3_ipaddress = False
+
 try:
     # Python 3.3+ or backport
-    from ipaddress import ip_address as _ip_address  # pylint: disable=F0401
+    from ipaddress import ip_network as _ip_address  # pylint: disable=F0401
```
Comment (vincentbernat) [inline lib/exabgp/application/healthcheck.py]: Could be done with ipaddr too:

```python
from ipaddr import IPNetwork as ip_address
```

**Comment by benagricola** [comment]
Hi @vincentbernat,

Cheers for pointing out that the `ipaddr` module supports this functionality too. I've removed the check requirements for the python3 module and merged this to provide the same functionality across both.

**Comment by vincentbernat** [comment]
Great! All good for me. Let's wait a bit if @thomas-mangin has something to say.

**Comment by thomas-mangin** [comment]
Yes I have a comment: I merged :-)

**Pull request merged**